optuna/study/_tell.py

```diff
@@ -97,106 +132,67 @@ def _tell_with_warning(
             Study.optimize.
     """
 
-    if not isinstance(trial, (trial_module.Trial, int)):
-        raise TypeError("Trial must be a trial object or trial number.")
-
-    if state == TrialState.COMPLETE:
-        if values is None:
-            raise ValueError(
-                "No values were told. Values are required when state is TrialState.COMPLETE."
-            )
-    elif state in (TrialState.PRUNED, TrialState.FAIL):
-        if values is not None:
-            raise ValueError(
-                "Values were told. Values cannot be specified when state is "
-                "TrialState.PRUNED or TrialState.FAIL."
-            )
-    elif state is not None:
-        raise ValueError(f"Cannot tell with state {state}.")
-
-    if isinstance(trial, trial_module.Trial):
-        trial_number = trial.number
-        trial_id = trial._trial_id
-    elif isinstance(trial, int):
-        trial_number = trial
-        try:
-            trial_id = study._storage.get_trial_id_from_study_id_trial_number(
-                study._study_id, trial_number
-            )
-        except NotImplementedError as e:
-            warnings.warn(
-                "Study.tell may be slow because the trial was represented by its number but "
-                f"the storage {study._storage.__class__.__name__} does not implement the "
-                "method required to map numbers back. Please provide the trial object "
-                "to avoid performance degradation."
-            )
-
-            trials = study.get_trials(deepcopy=False)
-
-            if len(trials) <= trial_number:
-                raise ValueError(
-                    f"Cannot tell for trial with number {trial_number} since it has not been "
-                    "created."
-                ) from e
-
-            trial_id = trials[trial_number]._trial_id
-        except KeyError as e:
-            raise ValueError(
-                f"Cannot tell for trial with number {trial_number} since it has not been "
-                "created."
-            ) from e
-    else:
-        assert False, "Should not reach."
-
-    frozen_trial = study._storage.get_trial(trial_id)
-    warning_message = None
-
+    frozen_trial = _get_frozen_trial(study, trial)
     if frozen_trial.state.is_finished() and skip_if_finished:
         _logger.info(
-            f"Skipped telling trial {trial_number} with values "
+            f"Skipped telling trial {frozen_trial.number} with values "
             f"{values} and state {state} since trial was already finished. "
             f"Finished trial has values {frozen_trial.values} and state {frozen_trial.state}."
         )
         return copy.deepcopy(frozen_trial)
-    elif frozen_trial.state == TrialState.WAITING:
-        raise ValueError("Cannot tell a waiting trial.")
+    elif frozen_trial.state != TrialState.RUNNING:
+        raise ValueError(f"Cannot tell a {frozen_trial.state.name} trial.")
```

Comment:
Originally, storage classes throw `RuntimeError` for finished trials. This PR changes to throw `ValueError` explicitly.